README.md
# Mezzio Skeleton and Installer

[![Build Status](https://travis-ci.com/mezzio/mezzio-skeleton.svg?branch=master)](https://travis-ci.com/mezzio/mezzio-skeleton)
[![Coverage Status](https://coveralls.io/repos/github/mezzio/mezzio-skeleton/badge.svg?branch=master)](https://coveralls.io/github/mezzio/mezzio-skeleton?branch=master)

*Begin developing PSR-15 middleware applications in seconds!*

[mezzio](https://github.com/mezzio/mezzio) builds on
[laminas-stratigility](https://github.com/laminas/laminas-stratigility) to
provide a minimalist PSR-15 middleware framework for PHP with routing, DI
container, optional templating, and optional error handling capabilities.

This installer will setup a skeleton application based on mezzio by
choosing optional packages based on user input as demonstrated in the following
screenshot:

![screenshot-installer](https://user-images.githubusercontent.com/1011217/90332191-55d32200-dfbb-11ea-80c0-27a07ef5691a.png)

The user selected packages are saved into `composer.json` so that everyone else
working on the project have the same packages installed. Configuration files and
templates are prepared for first use. The installer command is removed from
`composer.json` after setup succeeded, and all installer related files are
removed.

## Getting Started

Start your new Mezzio project with composer:

```bash
$ composer create-project mezzio/mezzio-skeleton <project-path>
```

After choosing and installing the packages you want, go to the
`<project-path>` and start PHP's built-in web server to verify installation:

```bash
$ composer run --timeout=0 serve
```

You can then browse to http://localhost:8080.

> ### Linux users
>
> On PHP versions prior to 7.1.14 and 7.2.2, this command might not work as
> expected due to a bug in PHP that only affects linux environments. In such
> scenarios, you will need to start the [built-in web
> server](http://php.net/manual/en/features.commandline.webserver.php) yourself,
> using the following command:
>
> ```bash
> $ php -S 0.0.0.0:8080 -t public/ public/index.php
> ```

> ### Setting a timeout
>
> Composer commands time out after 300 seconds (5 minutes). On Linux-based
> systems, the `php -S` command that `composer serve` spawns continues running
> as a background process, but on other systems halts when the timeout occurs.
>
> As such, we recommend running the `serve` script using a timeout. This can
> be done by using `composer run` to execute the `serve` script, with a
> `--timeout` option. When set to `0`, as in the previous example, no timeout
> will be used, and it will run until you cancel the process (usually via
> `Ctrl-C`). Alternately, you can specify a finite timeout; as an example,
> the following will extend the timeout to a full day:
>
> ```bash
> $ composer run --timeout=86400 serve
> ```

## Installing alternative packages

There is a feature to install alternative packages: Instead of entering one of
the selection **you can actually type the package name and version**.

> ```
>   Which template engine do you want to use?
>   [1] Plates
>   [2] Twig
>   [3] zend-view installs zend-servicemanager
>   [n] None of the above
>   Make your selection or type a composer package name and version (n): infw/pug:0.1
>   - Searching for infw/pug:0.1
>   - Adding package infw/pug (0.1)
> ```

That feature allows you to install any alternative package you want. It has its limitations though:

* The alternative package must follow this format `namespace/package:1.0`. It needs the correct version.
* Templates are not copied, but the ConfigProvider can be configured in such way that it uses the
  default templates directly from the package itself.
* This doesn't work for containers as the container.php file needs to be copied.

## Troubleshooting

If the installer fails during the ``composer create-project`` phase, please go
through the following list before opening a new issue. Most issues we have seen
so far can be solved by `self-update` and `clear-cache`.

1. Be sure to work with the latest version of composer by running `composer self-update`.
2. Try clearing Composer's cache by running `composer clear-cache`.

If neither of the above help, you might face more serious issues:

- Info about the [zlib_decode error](https://github.com/composer/composer/issues/4121).
- Info and solutions for [composer degraded mode](https://getcomposer.org/doc/articles/troubleshooting.md#degraded-mode).

## Application Development Mode Tool

This skeleton comes with [laminas-development-mode](https://github.com/laminas/laminas-development-mode).
It provides a composer script to allow you to enable and disable development mode.

### To enable development mode

**Note:** Do NOT run development mode on your production server!

```bash
$ composer development-enable
```

**Note:** Enabling development mode will also clear your configuration cache, to
allow safely updating dependencies and ensuring any new configuration is picked
up by your application.

### To disable development mode

```bash
$ composer development-disable
```

### Development mode status

```bash
$ composer development-status
```

## Configuration caching

By default, the skeleton will create a configuration cache in
`data/config-cache.php`. When in development mode, the configuration cache is
disabled, and switching in and out of development mode will remove the
configuration cache.

You may need to clear the configuration cache in production when deploying if
you deploy to the same directory. You may do so using the following:

```bash
$ composer clear-config-cache
```

You may also change the location of the configuration cache itself by editing
the `config/config.php` file and changing the `config_cache_path` entry of the
local `$cacheConfig` variable.

## Skeleton Development

This section applies only if you cloned this repo with `git clone`, not when you
installed mezzio with `composer create-project ...`.

If you want to run tests against the installer, you need to clone this repo and
setup all dependencies with composer.  Make sure you **prevent composer running
scripts** with `--no-scripts`, otherwise it will remove the installer and all
tests.

```bash
$ composer update --no-scripts
$ composer test
```

Please note that the installer tests remove installed config files and templates
before and after running the tests.

Before contributing read [the contributing guide](https://github.com/mezzio/.github/blob/master/CONTRIBUTING.md).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file, in reverse chronological order by release.

## 3.7.1 - TBD

### Added

- Nothing.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 3.7.0 - 2020-09-09


-----

### Release Notes for [3.7.0](https://github.com/mezzio/mezzio-skeleton/milestone/8)

Feature release (minor)

### 3.7.0

- Total issues resolved: **0**
- Total pull requests resolved: **2**
- Total contributors: **1**

#### Documentation

 - [35: Document installing alternative packages](https://github.com/mezzio/mezzio-skeleton/pull/35) thanks to @geerteltink

#### Enhancement

 - [34: Use router specific configuration](https://github.com/mezzio/mezzio-skeleton/pull/34) thanks to @geerteltink

## 3.6.0 - 2020-09-09


-----

### Release Notes for [3.6.0](https://github.com/mezzio/mezzio-skeleton/milestone/6)



### 3.6.0

- Total issues resolved: **2**
- Total pull requests resolved: **3**
- Total contributors: **3**

#### Enhancement

 - [33: Mark supported options in the installer](https://github.com/mezzio/mezzio-skeleton/pull/33) thanks to @geerteltink and @weierophinney
 - [32: Feat: Laminas coding style 2.1](https://github.com/mezzio/mezzio-skeleton/pull/32) thanks to @geerteltink

#### Bug

 - [31: Fix tests errors and drop PHP 7.2 support](https://github.com/mezzio/mezzio-skeleton/pull/31) thanks to @charlienux

## 3.5.0 - 2020-08-14

### Added

- [#25](https://github.com/mezzio/mezzio-skeleton/pull/25) adds support for chubbyphp/chubbyphp-container as a DI container option.

### Changed

- [#26](https://github.com/mezzio/mezzio-skeleton/pull/26) replaces the dependency on ocramius/package-versions with composer/package-versions-deprecated, which will allow immediate usage with laminas/laminas-cli.

### Deprecated

- Nothing.

### Removed

- [#25](https://github.com/mezzio/mezzio-skeleton/pull/25) drops support for PHP 7.1.

### Fixed

- Nothing.

## 3.4.1 - 2020-05-27

### Added

- Nothing.

### Changed

- [#24](https://github.com/mezzio/mezzio-skeleton/pull/24) updates the minimum versions of the various asset used, including Bootstrap, jQuery, and FontAwesome.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 3.4.0 - 2020-05-27

### Added

- [#19](https://github.com/mezzio/mezzio-skeleton/pull/19) adds support for Composer version 2 releases.

- [#17](https://github.com/mezzio/mezzio-skeleton/pull/17) adds support for Aura.Di 4.0.

### Changed

- [#20](https://github.com/mezzio/mezzio-skeleton/pull/20) updates the minimum supported version of Diactoros to 2.3.0, and registers the ConfigProvider for that component to ensure PSR-17 HTTP Message Factory implementations are exposed to the DI container.

- [#20](https://github.com/mezzio/mezzio-skeleton/pull/20) modifies the data returned by the `HomePageHandler` when returning a JSON response; it now includes containerName and routerName, and links to documentation for each.

- [#17](https://github.com/mezzio/mezzio-skeleton/pull/17) updates `laminas/laminas-auradi-config` dependency to version 2.0 in order to support Aura.Di 4.0. It is still compatible with Aura.Di 3.4.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 3.3.2 - 2020-04-21

### Added

- Nothing.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- [#15](https://github.com/mezzio/mezzio-skeleton/pull/15) fixes links to containers documentation.

## 3.3.1 - 2020-04-16

### Added

- Nothing.

### Changed

- [#14](https://github.com/mezzio/mezzio-skeleton/pull/14) updates asset: [jQuery 3.5.0](https://blog.jquery.com/2020/04/10/jquery-3-5-0-released/).

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 3.3.0 - 2020-03-20

### Added

- [zendframework/zend-expressive-skeleton#269](https://github.com/zendframework/zend-expressive-skeleton/pull/269) adds support for PHP 7.3.

- [#11](https://github.com/mezzio/mezzio-skeleton/pull/11) adds support for PHP 7.4.

### Changed

- [zendframework/zend-expressive-skeleton#268](https://github.com/zendframework/zend-expressive-skeleton/pull/268)
  and [#12](https://github.com/mezzio/mezzio-skeleton/pull/12) updates PHP-DI ContainerWrapper to version `^6.0`.

- [#10](https://github.com/mezzio/mezzio-skeleton/pull/10) updates assets:
  - jQuery 3.4.1
  - Bootstrap 4.4.1
  - FontAwesome 5.12.1

- [#11](https://github.com/mezzio/mezzio-skeleton/pull/11) and [#13](https://github.com/mezzio/mezzio-skeleton/pull/13) updates the following dependencies:
  - filp/whoops to `^2.7.1`
  - jsoumelidis/zend-sf-di-config to `^0.4`
  - laminas/laminas-auradi-config to `^1.0.2`
  - laminas/laminas-component-installer to `^2.1.2`
  - laminas/laminas-config-aggregator to `^1.2`
  - laminas/laminas-diactoros to `^1.8.7 || ^2.2.2`
  - laminas/laminas-pimple-config to `^1.1.1`
  - laminas/laminas-servicemanager to `^3.4`
  - mezzio/mezzio to `^3.2.1`
  - mezzio/mezzio-aurarouter to `^3.0.1`
  - mezzio/mezzio-fastroute to `^3.0.3`
  - mezzio/mezzio-helpers to `^5.3`
  - mezzio/mezzio-laminasrouter to `^3.0.1`
  - mezzio/mezzio-laminasviewrenderer to `^2.2`
  - mezzio/mezzio-platesrenderer to `^2.2`
  - mezzio/mezzio-tooling to `^1.3`
  - mezzio/mezzio-twigrenderer to `^2.6`
  - northwoods/container to `^3.2`

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- [#7](https://github.com/mezzio/mezzio-skeleton/pull/7) fixes path to favicon.

## 3.2.3 - 2018-11-05

### Added

- Nothing.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- [zendframework/zend-expressive-skeleton#266](https://github.com/zendframework/zend-expressive-skeleton/pull/266) adds check for PHP-DI ContainerWrapper to the homepage handler

## 3.2.2 - 2018-11-05

### Added

- Nothing.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed


src/MezzioInstaller/config.php
<?php

declare(strict_types=1);

// @phpcs:disable Generic.Files.LineLength.TooLong
return [
    'packages'    => [
        'chubbyphp/chubbyphp-laminas-config' => [
            'version' => '^1.0',
        ],
        'elie29/zend-phpdi-config'           => [
            'version' => '^6.0',
        ],
        'filp/whoops'                        => [
            'version' => '^2.7.1',
        ],
        'jsoumelidis/zend-sf-di-config'      => [
            'version' => '^0.4',
        ],
        'northwoods/container'               => [
            'version' => '^3.2',
        ],
        'laminas/laminas-auradi-config'      => [
            'version' => '^2.0.1',
        ],
        'mezzio/mezzio-aurarouter'           => [
            'version'   => '^3.0.1',
            'whitelist' => [
                'mezzio/mezzio-aurarouter',
            ],
        ],
        'mezzio/mezzio-fastroute'            => [
            'version'   => '^3.0.3',
            'whitelist' => [
                'mezzio/mezzio-fastroute',
            ],
        ],
        'mezzio/mezzio-platesrenderer'       => [
            'version'   => '^2.2',
            'whitelist' => [
                'mezzio/mezzio-platesrenderer',
            ],
        ],
        'mezzio/mezzio-twigrenderer'         => [
            'version'   => '^2.6',
            'whitelist' => [
                'mezzio/mezzio-twigrenderer',
            ],
        ],
        'mezzio/mezzio-laminasrouter'        => [
            'version'   => '^3.0.1',
            'whitelist' => [
                'mezzio/mezzio-laminasrouter',
            ],
        ],
        'mezzio/mezzio-laminasviewrenderer'  => [
            'version'   => '^2.2',
            'whitelist' => [
                'mezzio/mezzio-laminasviewrenderer',
            ],
        ],
        'laminas/laminas-pimple-config'      => [
            'version' => '^1.1.1',
        ],
        'laminas/laminas-servicemanager'     => [
            'version' => '^3.4',
        ],
    ],
    'require-dev' => [
        'filp/whoops',
    ],
    'application' => [
        'flat'    => [
            'packages'  => [],
            'resources' => [
                'Resources/src/ConfigProvider.flat.php' => 'src/App/ConfigProvider.php',
            ],
        ],
        'modular' => [
            'packages'  => [],
            'resources' => [
                'Resources/src/ConfigProvider.modular.php' => 'src/App/src/ConfigProvider.php',
            ],
        ],
    ],
    'questions'   => [
        'container'       => [
            'question'               => 'Which container do you want to use for dependency injection?',
            'default'                => 3,
            'required'               => true,
            'custom-package'         => true,
            'custom-package-warning' => 'You need to edit public/index.php to start the custom container.',
            'options'                => [
                1 => [
                    'name'     => 'Aura.Di <comment>(supported by laminas)</comment>',
                    'packages' => [
                        'laminas/laminas-auradi-config',
                    ],
                    'flat'     => [
                        'Resources/config/container-aura-di.php' => 'config/container.php',
                    ],
                    'modular'  => [
                        'Resources/config/container-aura-di.php' => 'config/container.php',
                    ],
                    'minimal'  => [
                        'Resources/config/container-aura-di.php' => 'config/container.php',
                    ],
                ],
                2 => [
                    'name'     => 'Pimple <comment>(supported by laminas)</comment>',
                    'packages' => [
                        'laminas/laminas-pimple-config',
                    ],
                    'flat'     => [
                        'Resources/config/container-pimple.php' => 'config/container.php',
                    ],
                    'modular'  => [
                        'Resources/config/container-pimple.php' => 'config/container.php',
                    ],
                    'minimal'  => [
                        'Resources/config/container-pimple.php' => 'config/container.php',
                    ],
                ],
                3 => [
                    'name'     => 'laminas-servicemanager <comment>(supported by laminas)</comment>',
                    'packages' => [
                        'laminas/laminas-servicemanager',
                    ],
                    'flat'     => [
                        'Resources/config/container-laminas-servicemanager.php' => 'config/container.php',
                    ],
                    'modular'  => [
                        'Resources/config/container-laminas-servicemanager.php' => 'config/container.php',
                    ],
                    'minimal'  => [
                        'Resources/config/container-laminas-servicemanager.php' => 'config/container.php',
                    ],
                ],
                4 => [
                    'name'     => 'Auryn',
                    'packages' => [
                        'northwoods/container',
                    ],
                    'flat'     => [
                        'Resources/config/container-auryn.php' => 'config/container.php',
                    ],
                    'modular'  => [
                        'Resources/config/container-auryn.php' => 'config/container.php',
                    ],
                    'minimal'  => [
                        'Resources/config/container-auryn.php' => 'config/container.php',
                    ],
                ],
                5 => [
                    'name'     => 'Symfony DI Container',
                    'packages' => [
                        'jsoumelidis/zend-sf-di-config',
                    ],
                    'flat'     => [
                        'Resources/config/container-sf-di.php' => 'config/container.php',
                    ],
                    'modular'  => [
                        'Resources/config/container-sf-di.php' => 'config/container.php',
                    ],
                    'minimal'  => [
                        'Resources/config/container-sf-di.php' => 'config/container.php',
                    ],
                ],
                6 => [
                    'name'     => 'PHP-DI',
                    'packages' => [
                        'elie29/zend-phpdi-config',
                    ],
                    'flat'     => [
                        'Resources/config/container-php-di.php' => 'config/container.php',
                    ],
                    'modular'  => [
                        'Resources/config/container-php-di.php' => 'config/container.php',
                    ],
                    'minimal'  => [
                        'Resources/config/container-php-di.php' => 'config/container.php',
                    ],
                ],
                7 => [
                    'name'     => 'chubbyphp-container',
                    'packages' => [
                        'chubbyphp/chubbyphp-laminas-config',
                    ],
                    'flat'     => [
                        'Resources/config/container-chubbyphp.php' => 'config/container.php',
                    ],
                    'modular'  => [
                        'Resources/config/container-chubbyphp.php' => 'config/container.php',
                    ],
                    'minimal'  => [
                        'Resources/config/container-chubbyphp.php' => 'config/container.php',
                    ],
                ],
            ],
        ],
        'router'          => [
            'question' => 'Which router do you want to use?',
            'default'  => 2,
            // TRUE: Must choose one / FALSE: May choose one or none of the above
            'required' => true,
            // Enable custom package input
            'custom-package' => true,
            // Display warning when choosing a custom package
            'custom-package-warning' => 'You need to write your own router adapter.',
            'options'                => [
                1 => [
                    'name'     => 'Aura.Router <comment>(supported by laminas)</comment>',
                    'packages' => [
                        'mezzio/mezzio-aurarouter',
                    ],
                    'flat'     => [
                        'Resources/config/routes-aura-router-full.php' => 'config/routes.php',
                    ],
                    'modular'  => [
                        'Resources/config/routes-aura-router-full.php' => 'config/routes.php',
                    ],
                    'minimal'  => [
                        'Resources/config/routes-aura-router-minimal.php' => 'config/routes.php',
                    ],
                ],
                2 => [
                    'name'     => 'FastRoute <comment>(supported by laminas)</comment>',
                    'packages' => [
                        'mezzio/mezzio-fastroute',
                    ],
                    'flat'     => [
                        'Resources/config/routes-fastroute-full.php' => 'config/routes.php',
                    ],
                    'modular'  => [
                        'Resources/config/routes-fastroute-full.php' => 'config/routes.php',
                    ],
                    'minimal'  => [
                        'Resources/config/routes-fastroute-minimal.php' => 'config/routes.php',
                    ],
                ],
                3 => [
                    'name'     => 'laminas-router <comment>(supported by laminas)</comment>',
                    'packages' => [
                        'mezzio/mezzio-laminasrouter',
                    ],
                    'flat'     => [
                        'Resources/config/routes-laminas-router-full.php' => 'config/routes.php',
                    ],
                    'modular'  => [
                        'Resources/config/routes-laminas-router-full.php' => 'config/routes.php',
                    ],
                    'minimal'  => [
                        'Resources/config/routes-laminas-router-minimal.php' => 'config/routes.php',
                    ],
                ],
            ],
        ],
        'template-engine' => [
            'question'       => 'Which template engine do you want to use?',
            'default'        => 'n',
            'required'       => false,
            'custom-package' => true,
            'options'        => [
                1 => [
                    'name'     => 'Plates <comment>(supported by laminas)</comment>',
                    'packages' => [
                        'mezzio/mezzio-platesrenderer',
                    ],
                    'flat'     => [
                        'Resources/templates/plates/404.phtml'       => 'templates/error/404.phtml',
                        'Resources/templates/plates/error.phtml'     => 'templates/error/error.phtml',
                        'Resources/templates/plates/layout.phtml'    => 'templates/layout/default.phtml',
                        'Resources/templates/plates/home-page.phtml' => 'templates/app/home-page.phtml',
                    ],
                    'modular'  => [
                        'Resources/templates/plates/404.phtml'       => 'src/App/templates/error/404.phtml',
                        'Resources/templates/plates/error.phtml'     => 'src/App/templates/error/error.phtml',
                        'Resources/templates/plates/layout.phtml'    => 'src/App/templates/layout/default.phtml',
                        'Resources/templates/plates/home-page.phtml' => 'src/App/templates/app/home-page.phtml',
                    ],
                    'minimal'  => [],
                ],
                2 => [
                    'name'     => 'Twig <comment>(supported by laminas)</comment>',
                    'packages' => [
                        'mezzio/mezzio-twigrenderer',
                    ],
                    'flat'     => [
                        'Resources/templates/twig/404.html.twig'       => 'templates/error/404.html.twig',
                        'Resources/templates/twig/error.html.twig'     => 'templates/error/error.html.twig',
                        'Resources/templates/twig/layout.html.twig'    => 'templates/layout/default.html.twig',
                        'Resources/templates/twig/home-page.html.twig' => 'templates/app/home-page.html.twig',
                    ],
                    'modular'  => [
                        'Resources/templates/twig/404.html.twig'       => 'src/App/templates/error/404.html.twig',
                        'Resources/templates/twig/error.html.twig'     => 'src/App/templates/error/error.html.twig',
                        'Resources/templates/twig/layout.html.twig'    => 'src/App/templates/layout/default.html.twig',
                        'Resources/templates/twig/home-page.html.twig' => 'src/App/templates/app/home-page.html.twig',
                    ],
                    'minimal'  => [],
                ],
                3 => [
                    'name'     => 'laminas-view <comment>installs laminas-servicemanager (supported by laminas)</comment>',
                    'packages' => [
                        'mezzio/mezzio-laminasviewrenderer',
                    ],
                    'flat'     => [
                        'Resources/templates/laminas-view/404.phtml'       => 'templates/error/404.phtml',
                        'Resources/templates/laminas-view/error.phtml'     => 'templates/error/error.phtml',
                        'Resources/templates/laminas-view/layout.phtml'    => 'templates/layout/default.phtml',
                        'Resources/templates/laminas-view/home-page.phtml' => 'templates/app/home-page.phtml',
                    ],
                    'modular'  => [
                        'Resources/templates/laminas-view/404.phtml'       => 'src/App/templates/error/404.phtml',
                        'Resources/templates/laminas-view/error.phtml'     => 'src/App/templates/error/error.phtml',
                        'Resources/templates/laminas-view/layout.phtml'    => 'src/App/templates/layout/default.phtml',
                        'Resources/templates/laminas-view/home-page.phtml' => 'src/App/templates/app/home-page.phtml',
                    ],
                    'minimal'  => [],
                ],
            ],
        ],
        'error-handler'   => [
            'question'       => 'Which error handler do you want to use during development?',
            'default'        => 1,
            'required'       => false,
            'custom-package' => true,
            'force'          => true,
            'options'        => [
                1 => [
                    'name'     => 'Whoops <comment>(supported by laminas)</comment>',
                    'packages' => [
                        'filp/whoops',
                    ],
                    'flat'     => [
                        'Resources/config/error-handler-whoops.php' => 'config/autoload/development.local.php.dist',
                    ],
                    'modular'  => [
                        'Resources/config/error-handler-whoops.php' => 'config/autoload/development.local.php.dist',
                    ],
                    'minimal'  => [
                        'Resources/config/error-handler-whoops.php' => 'config/autoload/development.local.php.dist',
                    ],
                ],
            ],
        ],
    ],
];
